app/models/spree/order.rb

```diff
@@ -618,6 +619,39 @@ def sorted_line_items
       end
     end
 
+    def can_generate_new_invoice?
+      return true if invoices.empty?
+
+      invoice_comparator.can_generate_new_invoice? current_state_invoice, invoices.last
```

Comment:
Epic work! :heart:

I had a little thought here about maybe reshaping the boundaries and responsibilities between the `Spree::Order` and `OrderInvoiceComparator` classes and modifying their public interface a bit.

So I think if it was me I would move all these new methods/logic out of `Spree::Order` and into `OrderInvoiceComparator` so that the `Spree::Order` class isn't holding the responsibility for knowing about the details of the invoice generating process. For example, the `#can_generate_invoice?` method in `OrderInvoiceComparator` could take an `order` as an argument, and then all the methods here like  `#current_state_invoice` and `#serialize_for_invoice` could be private methods inside `OrderInvoiceComparator`.

Then elsewhere where these methods need to be called like in the invoices controller or the order helper, instead of calling `order.can_generate_invoice?` it would be `OrderInvoiceComparator.new.can_generate_invoice?(order)` or the order could be an argument in the initializer, like `OrderInvoiceComparator.new(order).can_generate_invoice?`.

That class already has the responsibility of figuring out the answers to these questions, right?

It's not necessary right now, but it might be a nice refactor at some point. What do you think? :thinking:

Reply:
Something along these lines...

```ruby
class OrderInvoiceComparator
  def initializer(order)
    @order = order
  end

  def can_generate_new_invoice?
    return true if @order.invoices.empty?
    
    different?(current_state_invoice.presenter, latest_invoice.presenter,
               invoice_generation_selector)
  end
  
  ...
  
  private
  
  def current_state_invoice
      Invoice.new(
        order: @order,
        data: serialize_for_invoice,
        date: Time.zone.today,
        number: @order.invoices.count + 1
      )
  end
  
  def latest_invoice
    @order.invoices.last
  end
  
  def serialize_for_invoice
    Invoice::OrderSerializer.new(@order).serializable_hash
  end
  
  ...
  
```

Hmmm and actually some of these methods could be memoized so they don't do the same work twice in the same thread if both `#can_generate_invoice?` and `#can_update_latest_invoice?` get called on the same order

```ruby
  def serialize_for_invoice
    @serialize_for_invoice ||= Invoice::OrderSerializer.new(@order).serializable_hash
  end
```
```ruby
comparator = OrderInvoiceComparator.new(order)
comparator.can_generate_invoice?
comparator.can_update_latest_invoice?
```

:thinking::thinking::thinking:

Anyway, feel free to ignore my suggestions if you don't want to add any more to this PR (it's big already)!

Reply:
I like the idea, I'll update the code and request another review :)